pkg/compare/compare.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compare
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+
+	"sigs.k8s.io/yaml"
+
+	"istio.io/operator/pkg/tpath"
+)
+
+// YAMLCmpReporter is an option to generate tree based diff for YAMLs, used by cmp.Equal().
+type YAMLCmpReporter struct {
+	path     cmp.Path
+	diffTree map[string]interface{}
+}
+
+// PushStep implements Reporter.PushStep(PathStep) interface to keep track of current path by pushing
+// a step into YAMLCmpReporter.path
+func (r *YAMLCmpReporter) PushStep(ps cmp.PathStep) {
+	r.path = append(r.path, ps)
+}
+
+// PopStep implements Reporter.PopStep() interface to keep track of current path by popping a step out
+// of YAMLCmpReporter.path
+func (r *YAMLCmpReporter) PopStep() {
+	r.path = r.path[:len(r.path)-1]
+}
+
+// Report implements Reporter.Report(Result) to add diff path into YAMLCmpReporter.diffTree
+func (r *YAMLCmpReporter) Report(rs cmp.Result) {
+	if !rs.Equal() {
+		vx, vy := r.path.Last().Values()
+		var dm string
+		if vx.IsValid() && !vy.IsValid() {
+			dm = fmt.Sprintf("%v ->", vx)
+		} else if !vx.IsValid() && vy.IsValid() {
+			dm = fmt.Sprintf("-> %v", vy)
+		} else if vx.IsValid() && vy.IsValid() {
+			dm = fmt.Sprintf("%v -> %v", vx, vy)
+		}
+		if r.diffTree == nil {
+			r.diffTree = make(map[string]interface{})
+		}
+		_ = tpath.WriteNode(r.diffTree, pathToStringList(r.path), dm)
```

Comment:
why is it ok to ignore the error here?

Reply:
Great question! I do not see any chance we will get err by adding node into diffTree from existing trees.
I've updated this into panic(err) when err != nil rather than just ignoring it.